driver/src/main/java/org/neo4j/driver/internal/BaseDriver.java

```diff
@@ -67,33 +58,26 @@ public final Session session()
     @Override
     public final Session session( AccessMode mode )
     {
-        closedLock.readLock().lock();
-        try
+        assertOpen();
+        Session session = newSessionWithMode( mode );
+        if( closed.get() )
         {
-            assertOpen();
-            return newSessionWithMode( mode );
-        }
-        finally
-        {
-            closedLock.readLock().unlock();
+            // the driver is already closed and we either 1. obtain this session from the old session pool
+            // or 2. we obtain this session from a new session pool
+            // For 1. this closeResources will take no effect as everything is already closed.
+            // For 2. this closeResources will close the new connection pool just created to ensure no resource leak.
+            closeResources();
+            throw new IllegalStateException( "This driver instance has already been closed" );
```

Comment:
Maybe pull creation of this exception in a method so it is not duplicated here and in `assertOpen`